
--!nolint UnknownGlobal
--!nocheck


return function()
    local m = require(script.Parent.Parent.Helpers)
    describe("Helpers", function()
        describe("SetBit", function()
            it("Arithmetic", function()
                expect(m.SetBit(0, 0x00, true)).to.equal(0x01)
                expect(m.SetBit(0, 0x00, false)).to.equal(0x00)
                expect(m.SetBit(1, 0x00, true)).to.equal(0x02)
                expect(m.SetBit(1, 0x01, true)).to.equal(0x03)
            end)

            it("Range", function()
                expect(m.SetBit(8, 0xFF, true)).to.equal(0x01FF)
                expect(m.SetBit(-1, 0xFF, true)).to.throw()
            end)
        end)

        describe("SetFlag", function()
            it("Arithmetic", function()
                expect(m.SetFlag("C", 0x00, true)).to.equal(0x01)
                expect(m.SetFlag("Z", 0x00, true)).to.equal(0x02)
                expect(m.SetFlag("I", 0x00, true)).to.equal(0x04)
                expect(m.SetFlag("D", 0x00, true)).to.equal(0x08)
                expect(m.SetFlag("B", 0x00, true)).to.equal(0x10)
                expect(m.SetFlag("U", 0x00, true)).to.equal(0x20)
                expect(m.SetFlag("V", 0x00, true)).to.equal(0x40)
                expect(m.SetFlag("N", 0x00, true)).to.equal(0x80)
            end)

            it("Range", function()
                expect(function()
                    m.SetFlag("O", 0x00, true)
                end).to.throw()
                expect(function()
                    m.SetFlag("V", 0x00, true)
                end).never.to.throw()
            end)
        end)

        describe("GetBit", function()
            it("Arithmetic", function()
                expect(m.GetBit(0, 0x00)).to.equal(false)
                expect(m.GetBit(0, 0x01)).to.equal(true)
                expect(m.GetBit(1, 0x02)).to.equal(true)
                expect(m.GetBit(1, 0x01)).to.equal(false)
            end)

            it("Range", function()
                expect(m.GetBit(8, 0xFF)).to.equal(false)
                expect(m.GetBit(-1, 0xFF)).to.throw()
            end)
        end)

        describe("GetFlag", function()
            it("Arithmetic", function()
                expect(m.GetFlag("C", 0x01)).to.equal(true)
                expect(m.GetFlag("Z", 0x02)).to.equal(true)
                expect(m.GetFlag("I", 0x04)).to.equal(true)
                expect(m.GetFlag("D", 0x08)).to.equal(true)
                expect(m.GetFlag("B", 0x10)).to.equal(true)
                expect(m.GetFlag("U", 0x20)).to.equal(true)
                expect(m.GetFlag("V", 0x40)).to.equal(true)
                expect(m.GetFlag("N", 0x80)).to.equal(true)
                expect(m.GetFlag("Z", 0x01)).to.equal(false)
            end)

            it("Range", function()
                expect(function()
                    m.GetFlag("O", 0x00)
                end).to.throw()
                expect(function()
                    m.GetFlag("V", 0x00)
                end).never.to.throw()
            end)
        end)
    end)
end